src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexLexer.cs

```diff
@@ -0,0 +1,448 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Text.RegularExpressions;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.Common;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.VirtualChars;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.EmbeddedLanguages.RegularExpressions
+{
+    using static EmbeddedSyntaxHelpers;
+    using static RegexHelpers;
+
+    using RegexToken = EmbeddedSyntaxToken<RegexKind>;
+    using RegexTrivia = EmbeddedSyntaxTrivia<RegexKind>;
+
+    /// <summary>
+    /// Produces tokens from the sequence of <see cref="VirtualChar"/> characters.  Unlike the native C# and VB
+    /// lexer, this lexer is much more tightly controlled by the parser.  For example, while C# can have trivia
+    /// on virtual every token, the same is not true for RegexTokens.  As such, instead of automatically lexing
+    /// out tokens to make them available for the parser, the parser asks for each token as necessary passing
+    /// the right information to indicate which types and shapes of tokens are allowed.
+    /// 
+    /// The tight coupling means that the parser is allowed direct control of the position of the lexer.
+    /// 
+    /// Note: most of the time, tokens returned are just a single character long, including for long sequences
+    /// of text characters (like ```"goo"```).  This is just three <see cref="RegexTextNode"/>s in a row (each
+    /// containing a <see cref="RegexKind.TextToken"/> a single character long).
+    /// </summary>
+    internal struct RegexLexer
+    {
+        public readonly ImmutableArray<VirtualChar> Text;
+        public int Position;
+
+        public RegexLexer(ImmutableArray<VirtualChar> text) : this()
+        {
+            Text = text;
+        }
+
+        public VirtualChar CurrentChar => Position < Text.Length ? Text[Position] : new VirtualChar((char)0, default);
+
+        public ImmutableArray<VirtualChar> GetSubPattern(int start, int end)
```

Comment:
It looks like 90% of the calls to `GetSubPattern` use `Position` as their `end`. Could add an overload?

Reply:
I'll take a look.  Though my general preference is forcing callers to have to think about this stuff.  But maybe there could be "GetSubPatternToCurrentPosition" to make it clear.

Reply:
#resolved